README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

## Overview

`lorenzgini` is an [R](https://www.r-project.org) package that
calculates the Gini coefficient and plots Lorenz curves. It builds on
the `gini` function from package `DescTools`, and extends this function
to the generalized Gini as derived in:

> Bernasco, W. and W. Steenbeek (2017). More places than crimes:
> Implications for evaluating the law of crime concentration at place.
> *Journal of Quantitative Criminology*.
> <https://doi.org/10.1007/s10940-016-9324-7>

## Version

The most recent version of the package is:

> Steenbeek, W., Bernasco, W. (2018). *lorenzgini: generalized Gini for
> sparse data situations*. R package version 0.1.2. URL:
> <https://github.com/wsteenbeek/lorenzgini>

## Installation

You can install the package from this [GitHub
repository](https://github.com/wsteenbeek/lorenzgini). You first need to
install the [remotes](https://CRAN.R-project.org/package=remotes)
package.

``` r
install.packages("remotes")
```

Then install `lorenzgini` using the `install_github` function in the
[remotes](https://CRAN.R-project.org/package=remotes) package.

``` r
remotes::install_github("wsteenbeek/lorenzgini")
```

## How to start

After loading the package with `library(lorenzgini)`, see the
documentation of the two functions (these include examples) by typing:

  - `?gini`
  - `?lorenz`

## Vignette

The package includes a vignette explaining the background and the
functions in `lorenzgini` in more detail.

By far the easiest way to view the vignette is this direct link,
courtesy of the [GitHub HTML Preview
service](http://htmlpreview.github.io/):

  - [Introduction to
    lorenzgini](http://htmlpreview.github.io/?https://github.com/wsteenbeek/lorenzgini/blob/main/doc/lorenzgini.html)

If instead you want to access the vignettes from R itself you need to
take a few additional steps, because `remotes::install_github()` does
not build vignettes by default to save time and because it may require
additional packages.

1.  Install the rmarkdown package with `install.packages("rmarkdown")`

2.  [Install pandoc](http://johnmacfarlane.net/pandoc/installing.html)
    (and afterwards restart your computer)

3.  Then, install the package again but force building of the vignettes
    using `remotes::install_github("wsteenbeek/lorenzgini",
    build_vignettes = TRUE, force = TRUE)`. This will take a few
    minutes.

Afterwards, you should be able to view which vignettes are available
using:

``` r
browseVignettes("lorenzgini")
```

To directly read the vignettes rather than going through
`browseVignettes("lorenzgini")` you can use:

``` r
vignette("lorenzgini", package = "lorenzgini")
```

## License

This package is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License, version 3, as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose. See the GNU General
Public License for more details.

A copy of the GNU General Public License, version 3, is available at
<https://www.r-project.org/Licenses/GPL-3>


NEWS.md
# lorenzgini 0.1.2

* Updated to follow most recent installation convention using package 'remotes'.

# lorenzgini 0.1.1

* Updated `gini()` to add estimated Gini coefficient and bootstrapped confidence
  intervals indicators when outputting the kernel density plot of simulated
  Ginis.
  
* Updated 'lorenz()' function to only add the maximal line of equality and 'c/n'
  if Nunits > Nevents.

* Added vignette introducing the package and providing examples of `gini()` and
  `lorenz()`.

# lorenzgini 0.1.0

* First published version of the `lorenzgini` package. Includes functions
  'gini()' and 'lorenz()'.


R/lorenz.R
#' @import DescTools
#' @import ggplot2
NULL

#' Plots the (generalized) Lorenz Curve.
#'
#' This function plots a Lorenz curve. Following 'Bernasco, W. and W.
#' Steenbeek (2017). More places than crimes: Implications for evaluating the
#' law of crime concentration at place. Journal of Quantitative Criminology.
#' https://doi.org/10.1007/s10940-016-9324-7', if # events < # units of analysis,
#' vertical lines are added to the plot to indicate the percentage of units of
#' analysis where events could occur under maximum equality.
#'
#' @param mydata     a vector of N units, with counts of events per unit. The
#' vector should contain at least non-negative elements. The result will be NA,
#' if the vector contains negative elements.
#' @param prop         logical. Should the axes show proportions from 0 - 1
#' (default = TRUE), or percentages 0 - 100?
#' @param rescale      logical. Should the plot be rescaled so that 'c/n' is
#' moved to the end of the x-axis and the plot is stretched out accordingly?
#' @return             returns a Lorenz curve plot (generated by ggplot2)
#' @examples
#' # Plot a Lorenz curve of a non-sparse data situation:
#' # 10 units of analysis and 100 events
#' crimes <- c(22,18,17,12,11,8,6,3,2,1)
#' lorenz(crimes)
#' # Because sum(crimes) > length(crimes), it is the normal Lorenz curve.
#'
#' # Plot a Lorenz curve of a sparse data situation:
#' # 10 units of analysis and 5 events
#' crimes <- c(3,1,1,rep(0,7))
#' lorenz(crimes)
#'
#' # You can rescale the plot so that 'c/n' is moved to the end of the x-axis
#' # (and the graph is stretched to the right accordingly)
#' lorenz(crimes, rescale = TRUE)
#'
#' # Rescaling the graph is easier with very sparse data situation, as the
#' # adjusted line of maximal equality can become almost vertical from origin:
#' crimes <- c(3,1,1,rep(0,150))
#' lorenz(crimes)
#'
#' @export
lorenz <- function(mydata, prop = TRUE, rescale = FALSE){

  # from DescTools::Gini
  mydata <- as.numeric(mydata)
  if (any(is.na(mydata)) || any(mydata < 0)) return(NA_real_)

  mydata <- sort(mydata, decreasing = TRUE)

  if(rescale) {
    if(sum(mydata) < length(mydata)) mydata <- mydata[1:sum(mydata)]
  }

  mydata <- data.frame(events = mydata, idnr = seq_along(mydata))

  denom <- ifelse(prop, 100, 1)

  # fill data.frame
  mydata$ones <- 1
  mydata$percEvents <- mydata$events / sum(mydata$events) * 100
  mydata$cumpercEvents <- cumsum(mydata$percEvents)
  mydata$cumpercUnit <- cumsum(mydata$ones) / nrow(mydata) * 100

  Nunits <- nrow(mydata)
  Nevents <- sum(mydata$events)
  NunitsWithEvents <- nrow(mydata[mydata$events>0,])

  # This block is to create alternative variables with which to create adjusted plot
  mydata$cumpercUnit.new <- cumsum(mydata$ones) / Nevents * 100
  mydata$cumpercUnit.new[mydata$cumpercUnit.new>100] <- 100

  # take subset of data, to easily draw lines for generalized Gini later
  mydata_gen <- mydata[1:min(Nevents, nrow(mydata)),]
  # add first line with zero's, to make origin of curves start at 0,0
  mydata_gen <- rbind(c(rep(NA,4),0,0,0,0,0), mydata_gen)
  # where to draw the 'c/n' line?
  xaxis.point <- max(mydata_gen$cumpercUnit)

  # add first line with zero's, to make origin of curves start at 0,0
  mydata <- rbind(c(rep(NA,4),0,0,0), mydata)

  # prepare plot
  myplot <- ggplot2::ggplot(data=mydata, aes(x=cumpercUnit/denom, y=cumpercEvents/denom)) +
    ylab(paste("Cum.", ifelse(prop, "Prop.", "Percent.") , "Events")) + xlab(paste("Cum.", ifelse(prop, "Prop.", "Percent.") , "Units of Analysis")) +
    theme_light() +
    theme(legend.position="none") +
    scale_y_continuous(breaks=seq(0,100,20)/denom) +
    scale_x_continuous(breaks=seq(0,100,20)/denom) +
    geom_segment(aes(x=0, xend=100/denom, y=0, yend=100/denom), linetype="longdash") +
    geom_line()

  # add new line of maximal equality if Nunits > Nevents
  if (Nunits > Nevents){
    myplot <- myplot +
      geom_segment(data = mydata_gen, aes(x=0, xend=max(cumpercUnit/denom), y=0, yend=100/denom), linetype=3) +
      geom_segment(data = mydata_gen, aes(x=max(cumpercUnit/denom), xend=max(cumpercUnit/denom), y=2/denom, yend=100/denom), linetype=3) +
      annotate("text", x = xaxis.point/denom, y = 0, label = "c/n", size=4)
  }

  return(myplot)
}


R/gini.R
#' @import boot
#' @import DescTools
#' @import ggplot2
NULL

#' Calculates the (generalized) Gini coefficient
#'
#' This function calculates the Gini coefficient. It builds on the
#' `gini` function from package `DescTools`, and extends this
#' function to the generalized Gini.
#'
#' See 'Bernasco, W. and W. Steenbeek (2017). More places than crimes:
#' Implications for evaluating the law of crime concentration at place.
#' Journal of Quantitative Criminology. https://doi.org/10.1007/s10940-016-9324-7'
#' for more information.
#'
#' @param mydata     a vector of N units, with counts of events per unit. The
#' vector should contain at least non-negative elements. The result will be NA,
#' if the vector contains negative elements.
#' @param generalized  logical. Should the generalized gini be calculated?
#' (default = TRUE)
#' @param unbiased     logical. Should the unbiased Gini be calculated?
#' (default = TRUE)
#' @param conf.level   confidence level for the confidence interval, restricted
#' to lie between 0 and 1. If set, bootstrap confidence intervals are
#' calculated. If set to NA (default) no confidence intervals are returned.
#' @param R            number of bootstrap replicates. Usually this will be a
#' single positive integer. For importance resampling, some resamples may use
#' one set of weights and others use a different set of weights. In this case
#' R would be a vector of integers where each component gives the number of
#' resamples from each of the rows of weights. This is ignored if no confidence
#' intervals are to be calculated.
#' @param type         character string representing the type of interval
#' required. The value should be one out of the c("norm","basic", "stud",
#' "perc" or "bca"). This argument is ignored if no confidence intervals are
#' to be calculated.
#' @param plot         logical. Should a kernel density plot of the distribution
#' of R Ginis be printed? (default = FALSE). This argument is ignored if no
#' confidence intervals are to be calculated.
#' @return             If conf.level is set to NA then the result will be a
#' single numeric value and if a conf.level is provided, a named numeric vector
#' with 3 elements: gini (Gini coefficient), lwr.ci	(lower bound of the
#' confidence interval), upr.ci	(upper bound of the confidence interval)
#' @examples
#'
#' crimes <- c(10,8,2,0,0,rep(0,5))
#' crimes
#' length(crimes) # 10 units of analysis
#' sum(crimes) # 20 crimes in total
#' # Standard Gini:
#' gini(crimes, generalized = FALSE, unbiased = FALSE) # 0.78
#' # Is the same the generalized Gini, as Ncrimes >= Nunits:
#' gini(crimes, generalized = TRUE, unbiased = FALSE) # 0.78
#'
#' crimes <- c(5,2,1,0,0,rep(0,5))
#' crimes
#' length(crimes) # 10 units of analysis
#' sum(crimes) # 8 crimes in total
#' # Standard Gini:
#' gini(crimes, generalized = FALSE, unbiased = FALSE) # 0.8
#' # Is not the same the generalized Gini, as Ncrimes < Nunits:
#' gini(crimes, generalized = TRUE, unbiased = FALSE) # 0.75
#'
#' # Example of unbiased standard Gini:
#' crimes <- c(10, rep(0,9))
#' crimes
#' length(crimes) # 10 units of analysis
#' sum(crimes) # 10 crimes in total
#' gini(crimes, generalized = FALSE, unbiased = FALSE) # 0.9
#' # All crime is concentrated in one unit, but the Gini does not equal 1.
#'
#' crimes <- c(10, rep(0,99))
#' crimes
#' gini(crimes, generalized = FALSE, unbiased = FALSE) # 0.99
#'
#' # As the number of units with 0 crimes inceases, the Gini asymptotically
#' # approaches 1. This is adjusted in the 'unbiased' Gini, the default in `DescTools`:
#' crimes <- c(10, rep(0,9))
#' gini(crimes, generalized = FALSE, unbiased = TRUE) # 1
#' crimes <- c(10, rep(0,99))
#' gini(crimes, generalized = FALSE, unbiased = TRUE) # 1
#'
#' # generate crime events from zero-inflated poisson disribution
#' set.seed(8722)
#' crimes <- ifelse(rbinom(1000, size = 1, prob = .1) > 0, 0, rpois(1000, lambda = 1))
#'
#' # frequency of crimes
#' table(crimes)
#'
#' # Bootstrap-based confidence intervals:
#' set.seed(347)
#' gini(crimes, conf.level = .95)
#' # Also output kernel density plot of generated Ginis, with estimated Gini
#' # indicated as solid red line, and the confidence intervals delineated
#' # by dashed red lines:
#' set.seed(347)
#' gini(crimes, conf.level = .95, plot = TRUE)
#' # increasing number of simulations
#' set.seed(347)
#' gini(crimes, conf.level = .95, R = 20000, plot = TRUE)
#'
#' @export
gini <- function(mydata, generalized = TRUE, unbiased = TRUE, conf.level = NA, R = 1000, type = "bca", plot = FALSE){

  # from DescTools::Gini
  # cast to numeric, as else sum(mydata * 1:n) might overflow for integers
  # http://stackoverflow.com/questions/39579029/integer-overflow-error-using-gini-function-of-package-desctools
  mydata <- as.numeric(mydata)

  # from DescTools::Gini
  if (any(is.na(mydata)) || any(mydata < 0)) return(NA_real_)

  mygini <- function(x, generalized = TRUE, unbiased = TRUE){

    Nunits <- length(x)
    Ncrimes <- sum(x)

    # if we don't want the Gini for sparse data, simply
    # run the DescTools::Gini function
    if(!generalized){
      output <- DescTools::Gini(x, unbiased = unbiased)
    }

    # for Gini for sparse data:
    if(generalized){

      # return biased generalized Gini
      if(!unbiased){
        Gini.biased <- DescTools::Gini(x, unbiased = unbiased)
        output <- max(1, Nunits/Ncrimes) * (Gini.biased - 1) + 1
      }

      # return unbiased Gini for sparse data
      if(unbiased){
        Gini.biased <- DescTools::Gini(x, unbiased = FALSE)
        Gini.gen.biased <- max(1, Nunits/Ncrimes) * (Gini.biased - 1) + 1
        Gini.gen.unbiased <- Gini.gen.biased * min(Nunits, Ncrimes) / (min(Nunits, Ncrimes) - 1)
        output <- Gini.gen.unbiased
      }
    }

    return(output)

  }

  if(is.na(conf.level)){
    res <- mygini(mydata, generalized = generalized, unbiased = unbiased)
  } else {
    # adjusted bootstrap percentile (BCa) interval
    boot.gini <- boot::boot(mydata, function(x, d) mygini(x[d], generalized = generalized, unbiased = unbiased), R = R)
    ci <- boot::boot.ci(boot.gini, conf = conf.level, type = type)

    if(plot) {
      # how much should x-axis be shown left from mean and right from mean?
      dev <- max(abs(range(boot.gini$t) - boot.gini$t0))
      xlimits <- c(boot.gini$t0 - dev, boot.gini$t0 + dev)

      myplot <- ggplot2::ggplot(data = data.frame(id = seq_along(boot.gini$t), gini = boot.gini$t), aes(x = gini)) +
        geom_density(fill = "#4271AE", colour = "#1F3552",
                     alpha = 0.6) +
        scale_x_continuous(name = paste0("N = ", length(boot.gini$t))) +
        scale_y_continuous(name = "Density") +
        geom_vline(xintercept = boot.gini$t0, linetype="solid", color = "red") +
        geom_vline(xintercept = ci[[4]][4], linetype="dashed", color = "red") +
        geom_vline(xintercept = ci[[4]][5], linetype="dashed", color = "red") +
        expand_limits(x = xlimits) +
        ggtitle("Kernel density plot of distribution of Ginis")

      print(myplot)
    }

    res <- c(gini = boot.gini$t0, lwr.ci = ci[[4]][4], upr.ci = ci[[4]][5])
  }

  return(res)
}
